File: fcfs_scheduler.py
from .scheduler import Scheduler
from .process import Process
from typing import List

class FCFSScheduler(Scheduler):
    """First Come First Serve Scheduler Implementation"""
    
    def schedule(self) -> None:
        # Sort processes by arrival time
        self.processes.sort(key=lambda x: x.arrival_time)
        
        self.current_time = 0
        self.execution_sequence = []
        
        for process in self.processes:
            # If there's a gap between processes, move current time to process arrival
            if self.current_time < process.arrival_time:
                self.current_time = process.arrival_time
            
            # Set process start time
            process.start_time = self.current_time
            
            # Execute process
            start = self.current_time
            self.current_time += process.burst_time
            
            # Set completion time
            process.completion_time = self.current_time
            
            # Record execution sequence for visualization
            self.execution_sequence.append((process, start, self.current_time))

File: FCFSScheduler.java
package com.cpuscheduler.core;

import com.cpuscheduler.model.Process;
import java.util.*;

/**
 * First-Come, First-Served (FCFS) CPU scheduling algorithm.
 * Processes are executed in the order they arrive.
 */
public class FCFSScheduler extends AbstractScheduler {

    @Override
    public String getName() {
        return "First-Come, First-Served (FCFS)";
    }

    @Override
    public List<String> execute() {
        // Reset state
        ganttChart.clear();
        completedProcesses.clear();
        currentTime = 0;
        contextSwitches = 0;

        // Sort processes by arrival time
        sortByArrivalTime();
        Queue<Process> readyQueue = new LinkedList<>(processes);

        while (!allProcessesCompleted()) {
            // Get the next process to execute
            Process currentProcess = null;
            for (Process p : readyQueue) {
                if (p.getArrivalTime() <= currentTime && !p.isCompleted()) {
                    currentProcess = p;
                    break;
                }
            }

            if (currentProcess == null) {
                // No process is ready to execute, increment time
                currentTime++;
                continue;
            }

            // Execute the process until completion
            int startTime = currentTime;
            int executionTime = currentProcess.getBurstTime();
            currentTime += executionTime;
            
            // Update process completion
            currentProcess.execute(executionTime, startTime);
            completedProcesses.add(currentProcess);
            
            // Update Gantt chart
            updateGanttChart(currentProcess, startTime, currentTime);
            
            // Count context switch (only if there are more processes to come)
            if (!allProcessesCompleted()) {
                contextSwitches++;
            }
        }

        return ganttChart;
    }
}
